### PR TITLE
Add table azure_data_factory_pipeline. Closes #151

### LINKED ISSUE #151: Add table azure_datafactory_pipeline
**References**
https://docs.microsoft.com/en-us/azure/data-factory/quickstart-create-data-factory-rest-api

### PR DESCRIPTION
**Pull request by rajeshbal65**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT 300

SETUP: tests/azure_data_factory_pipeline []

PRETEST: tests/azure_data_factory_pipeline

TEST: tests/azure_data_factory_pipeline
Running terraform
data.azurerm_client_config.current: Refreshing state...
azurerm_resource_group.named_test_resource: Creating...
azurerm_resource_group.named_test_resource: Creation complete after 3s [id=/subscriptions/********-****-****-****-************/resourceGroups/turbottest38586]
azurerm_data_factory.named_test_resource: Creating...
azurerm_data_factory.named_test_resource: Still creating... [10s elapsed]
azurerm_data_factory.named_test_resource: Still creating... [20s elapsed]
azurerm_data_factory.named_test_resource: Creation complete after 22s [id=/subscriptions/********-****-****-****-************/resourceGroups/turbottest38586/providers/Microsoft.DataFactory/factories/turbottest38586]
azurerm_data_factory_pipeline.named_test_resource: Creating...
azurerm_data_factory_pipeline.named_test_resource: Creation complete after 4s [id=/subscriptions/********-****-****-****-************/resourceGroups/turbottest38586/providers/Microsoft.DataFactory/factories/turbottest38586/pipelines/turbottest38586]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 3 added, 0 changed, 3 destroyed.

Outputs:

resource_aka = azure:///subscriptions/********-****-****-****-************/resourceGroups/turbottest38586/providers/Microsoft.DataFactory/factories/turbottest38586/pipelines/turbottest38586
resource_aka_lower = azure:///subscriptions/********-****-****-****-************/resourcegroups/turbottest38586/providers/microsoft.datafactory/factories/turbottest38586/pipelines/turbottest38586
resource_id = /subscriptions/********-****-****-****-************/resourceGroups/turbottest38586/providers/Microsoft.DataFactory/factories/turbottest38586/pipelines/turbottest38586
resource_name = turbottest38586
subscription_id = ********-****-****-****-************

Running SQL query: test-get-query.sql
[
  {
    "id": "/subscriptions/********-****-****-****-************/resourceGroups/turbottest38586/providers/Microsoft.DataFactory/factories/turbottest38586/pipelines/turbottest38586",
    "name": "turbottest38586",
    "resource_group": "turbottest38586",
    "subscription_id": "********-****-****-****-************"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "id": "/subscriptions/********-****-****-****-************/resourceGroups/turbottest38586/providers/Microsoft.DataFactory/factories/turbottest38586/pipelines/turbottest38586",
    "name": "turbottest38586",
    "resource_group": "turbottest38586",
    "subscription_id": "********-****-****-****-************"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "azure:///subscriptions/********-****-****-****-************/resourceGroups/turbottest38586/providers/Microsoft.DataFactory/factories/turbottest38586/pipelines/turbottest38586",
      "azure:///subscriptions/********-****-****-****-************/resourcegroups/turbottest38586/providers/microsoft.datafactory/factories/turbottest38586/pipelines/turbottest38586"
    ],
    "name": "turbottest38586",
    "title": "turbottest38586"
  }
]
✔ PASSED

POSTTEST: tests/azure_data_factory_pipeline

TEARDOWN: tests/azure_data_factory_pipeline

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
### Basic info

```sql
select
  name,
  id,
  type,
  etag,
  factory_name
from
  azure_data_factory_pipeline;

+-----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------+--------------------------------------+-----------------+
| name            | id                                                                                                                                                                     | type                                      | etag                                 | factory_name    |
+-----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------+--------------------------------------+-----------------+
| turbottest44547 | /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest44547/providers/Microsoft.DataFactory/factories/turbottest44547/pipelines/turbottest44547 | Microsoft.DataFactory/factories/pipelines | 070074c7-0000-0100-0000-60e2d8650000 | turbottest44547 |
+-----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------+--------------------------------------+-----------------+
```
```
select
  p.name,
  p.id,
  p.factory_name,
  f.encryption factory_encryption
from
  azure_data_factory_pipeline p,
  azure_data_factory f
  where f.encryption is null;
+-----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------+--------------------+
| name            | id                                                                                                                                                                     | factory_name    | factory_encryption |
+-----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------+--------------------+
| turbottest35948 | /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest35948/providers/Microsoft.DataFactory/factories/turbottest35948/pipelines/turbottest35948 | turbottest35948 | <null>             |
+-----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------+--------------------+
```
</details>

### CODE REVIEW

File: azure/table_azure_data_factory.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/datafactory/mgmt/2018-06-01/datafactory"
+	"github.com/turbot/steampipe-plugin-sdk/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+)
+
+//// TABLE DEFINITION ////
```
Comment (Subhajit97) [inline azure/table_azure_data_factory.go:13]:
```suggestion
//// TABLE DEFINITION
```

File: azure/table_azure_data_factory_pipeline.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/datafactory/mgmt/2018-06-01/datafactory"
+	"github.com/turbot/steampipe-plugin-sdk/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+)
+
+//// TABLE DEFINITION ////
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:14]:
```suggestion
//// TABLE DEFINITION
```

File: azure/table_azure_data_factory_pipeline.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+func tableAzureDataFactoryPipeline(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "azure_data_factory_pipeline",
+		Description: "Azure Data Factory Pipeline",
+		Get: &plugin.GetConfig{
+			KeyColumns:        plugin.AllColumns([]string{"name", "resource_group", "factory_name"}),
+			Hydrate:           getPipeline,
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:22]:
```suggestion
			Hydrate:           getDataFactoryPipeline,
```

File: azure/table_azure_data_factory_pipeline.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+			ShouldIgnoreError: isNotFoundError([]string{"ResourceNotFound", "ResourceGroupNotFound", "404"}),
+		},
+		List: &plugin.ListConfig{
+			Hydrate:       listPipelines,
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:26]:
```suggestion
			Hydrate:       listDataFactoryPipelines,
```

File: azure/table_azure_data_factory_pipeline.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Etag"),
+			},
+			{
+				Name:        "type",
+				Description: "The resource type.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "factory_name",
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:59]: Add this column after `id`

File: azure/table_azure_data_factory_pipeline.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+				Description: "Time the factory was created in ISO8601 format.",
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:60]: Fix description

File: azure/table_azure_data_factory_pipeline.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("Pipeline.RunDimensions"),
+			},
+			{
+				Name:        "pipeline_policy",
+				Description: "Pipeline ElapsedTime Metric Policy.",
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("Pipeline.Folder.PipelinePolicy"),
+			},
+			// Standard columns
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:111]:
```suggestion
			},

			// Steampipe standard columns
```

File: azure/table_azure_data_factory_pipeline.go [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Name"),
+			},
+			{
+				Name:        "akas",
+				Description: ColumnDescriptionAkas,
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("ID").Transform(idToAkas),
+			},
+			{
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:124]:
```suggestion
			},

			// Azure standard columns
			{
```

File: azure/table_azure_data_factory_pipeline.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+		},
+	}
+}
+
+type pipelineInfo = struct {
+	datafactory.PipelineResource
+	FactoryName string
+}
+
+//// LIST FUNCTIONS ////
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:145]:
```suggestion
//// LIST FUNCTION
```

File: azure/table_azure_data_factory_pipeline.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+	for pagesLeft {
+		result, err := pipelineClient.ListByFactory(ctx, resourceGroup, *factoryInfo.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, pipeline := range result.Values() {
+			// plugin.Logger(ctx).Trace("getPipeline1111", pipeline)
+			d.StreamListItem(ctx, pipelineInfo{pipeline, *factoryInfo.Name})
+			// d.StreamListItem(ctx, pipeline)
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:171]: Remove unused codes
```suggestion
			d.StreamListItem(ctx, pipelineInfo{pipeline, *factoryInfo.Name})
```

File: azure/table_azure_data_factory_pipeline.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+		}
+		result.NextWithContext(context.Background())
+		pagesLeft = result.NotDone()
+	}
+	return nil, err
+}
+
+//// HYDRATE FUNCTIONS ////
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:179]:
```suggestion
//// HYDRATE FUNCTIONS
```

File: azure/table_azure_data_factory_pipeline.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+
+func getPipeline(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getPipeline")
+
+	pipelineName := d.KeyColumnQuals["name"].GetStringValue()
+	resourceGroup := d.KeyColumnQuals["resource_group"].GetStringValue()
+	factoryName := d.KeyColumnQuals["factory_name"].GetStringValue()
+
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:187]: Add check for empty input

File: azure/table_azure_data_factory_pipeline.go [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+				Name:        "description",
+				Description: "The description of the pipeline.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Pipeline.Description"),
+			},
+			{
+				Name:        "etag",
+				Description: "Etag identifies change in the resource.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Etag"),
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:56]: Not required
```suggestion
```

File: azure/table_azure_data_factory_pipeline.go [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+			{
+				Name:        "variables",
+				Description: "List of variables for pipeline.",
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("Pipeline.Variables"),
+			},
+			{
+				Name:        "concurrency",
+				Description: "The max number of concurrent runs for the pipeline.",
+				Type:        proto.ColumnType_JSON,
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:90]: Is the type correct? 
Seems like the value is of type `integer`

File: azure/table_azure_data_factory_pipeline.go [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+		return nil, err
+	}
+
+	// In some cases resource does not give any notFound error
+	// instead of notFound error, it returns empty data
+	if op.ID != nil {
+		return pipelineInfo{op, factoryName}, nil
+	}
+
+	return op, nil
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_pipeline.go:213]:
```suggestion
	return nil, nil
```

**Review by Subhajit97 — APPROVED** [review]

**Review by rajlearner17 — CHANGES_REQUESTED** [review]
Let's add more example to check the Data Factory version in use or add the same to the example query as additional column. 
Query to check if encryption using CMK or not.

**Review by rajlearner17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**